lib/marked.js

```diff
@@ -554,9 +554,51 @@ inline.normal = merge({}, inline);
 inline.pedantic = merge({}, inline.normal, {
   strong: /^__(?=\S)([\s\S]*?\S)__(?!_)|^\*\*(?=\S)([\s\S]*?\S)\*\*(?!\*)/,
   em: /^_(?=\S)([\s\S]*?\S)_(?!_)|^\*(?=\S)([\s\S]*?\S)\*(?!\*)/,
-  link: edit(/^!?\[(label)\]\(\s*<?([\s\S]*?)>?(?:\s+(['"][\s\S]*?['"]))?\s*\)/)
-    .replace('label', inline._label)
-    .getRegex(),
+  /* Original link re: /^!?\[(label)\]\(\s*<?([\s\S]*?)>?(?:\s+(['"][\s\S]*?['"]))?\s*\)/
+   * This captures the spec reasonably well but is vulnerable to REDOS.
+   * Instead we use a custom parser that follows the RegExp.exec semantics. */
+  link: {
+    exec: function (s) {
+      // [TEXT](DESTINATION)
+      var generalLinkRe = edit(/^!?\[(label)\]\((.*?)\)/)
+        .replace('label', inline._label)
+        .getRegex();
+
+      function unwrapAngleBrackets (str) {
+        if (str.match(/^<.*>$/)) {
+          str = str.slice(1, -1);
+        }
+        return str;
+      }
+
+      var fullMatch = generalLinkRe.exec(s);
+      if (fullMatch) {
+        var text = fullMatch[1];
+        var destination = fullMatch[2];
+
+        var m;
+
+        var destinationAndTitleRe = /^([^'"(]*[^\s])\s+(['"(].*['")])/;
+        if (m = destinationAndTitleRe.exec(destination)) {
+          // <destination> -> destination
+          var dest1 = m[1].trim();
+          dest1 = unwrapAngleBrackets(dest1);
+          var title1 = m[2];
+          return [fullMatch[0], text, dest1, title1];
+        }
+
+        var destinationRe = /^(<?[\s\S]*>?)/;
+        if (m = destinationRe.exec(destination)) {
```

Comment:
The two `if blocks` are nearly identical. Can you make a functions for those?
```

Reply:
Agree, but not a deal breaker for my review.

Reply:
@styfle See ba2fc13